### PR TITLE
fix export class declare with name object targeting es5 (#17494)

### LINKED ISSUE #17494: Module with "Object" export has bad ES5 emit
**TypeScript Version:** 2.2 to current

**Code**
TS file:
`export class Object {}`
tsconfig.json:
`{"compilerOptions": {"target": "es5"}}`

**Expected behavior:**
The module should be able to initialize like it did in TypeScript 2.1.

**Actual behavior:**
Module fails to initialize; `TypeError: Object is undefined`.

The emitted code is
```
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Object = (function () {
    function Object() {
    }
    return Object;
}());
exports.Object = Object;
```
Clearly `Object.defineProperty` is intended to reference the global `Object`, but it gets an uninitialized local variable instead.

### PR DESCRIPTION
**Pull request by Kingwl**
Fixes #17494

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
A few things here:

1. We should definitely not error when this occurs in an ambient context.
2. I'm pretty sure we shouldn't error on this if your *module target* is ES2015 or ESNext.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -21636,7 +21636,20 @@ namespace ts {
...
-                    error(name, message, (<Identifier>name).escapedText as string);
+                    error(name, message, name.escapedText as string);
+            }
+        }
+
+        /**
+         * The name cannot be used as names of user defined types with special target.
+         */
+        function checkTypeNameIsDenyWithTarget(name: Identifier, message: DiagnosticMessage): void {
+            const target = compilerOptions.target || ScriptTarget.ES3;
```
Comment (mhegazy) [inline src/compiler/checker.ts]: already defined in `languageVersion`

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -21636,7 +21636,20 @@ namespace ts {
...
+            if (target === ScriptTarget.ES5) {
+                switch (name.escapedText) {
```
Comment (mhegazy) [inline src/compiler/checker.ts]: use `if` here.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -21636,7 +21636,20 @@ namespace ts {
...
+                    case "Object": // https://github.com/Microsoft/TypeScript/issues/17494
```
Comment (mhegazy) [inline src/compiler/checker.ts]: consider:

```ts
if (languageVersion === ScriptTarget.ES5 && name.escapedText === "Object") {
    error(name, Diagnostics.Class_name_cannot_be_0_when_targeting_1, name.escapedText as string, "ES5");
}
```

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -21761,6 +21774,7 @@ namespace ts {
             checkDecorators(node);
             if (node.name) {
                 checkTypeNameIsReserved(node.name, Diagnostics.Class_name_cannot_be_0);
+                checkTypeNameIsDenyWithTarget(node.name, Diagnostics.Class_name_cannot_be_0_when_targeting_1);
```
Comment (mhegazy) [inline src/compiler/checker.ts]: move this lower to be under ` if (!isInAmbientContext(node)) {` block

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -21636,7 +21636,20 @@ namespace ts {
...
                 case "object":
-                    error(name, message, (<Identifier>name).escapedText as string);
+                    error(name, message, name.escapedText as string);
+            }
+        }
+
+        /**
+         * The name cannot be used as names of user defined types with special target.
+         */
+        function checkTypeNameIsDenyWithTarget(name: Identifier, message: DiagnosticMessage): void {
```
Comment (mhegazy) [inline src/compiler/checker.ts]: consider renaming to `checkClassNameCollisionWithObject`

File: tests/cases/conformance/es6/modules/exportClassWithDenyName.ts [hunk 6]
```diff
@@ -0,0 +1,2 @@
+// @target: ES5
+export class Object {}
```
Comment (mhegazy) [inline tests/cases/conformance/es6/modules/exportClassWithDenyName.ts]: also add:
- class expression
- export default class
- ambient class (should not error)

**Comment by mhegazy** [comment]
@Kingwl mind refreshing this PR

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -24279,8 +24289,12 @@ namespace ts {
                 checkTypeNameIsReserved(node.name, Diagnostics.Class_name_cannot_be_0);
                 checkCollisionWithRequireExportsInGeneratedCode(node, node.name);
                 checkCollisionWithGlobalPromiseInGeneratedCode(node, node.name);
+                if (!(node.flags & NodeFlags.Ambient)) {
+                    checkClassNameCollisionWithObject(node.name);
+                }
             }
             checkTypeParameters(getEffectiveTypeParameterDeclarations(node));
+            checkTypeParameters(node.typeParameters);
```
Comment (mhegazy) [inline src/compiler/checker.ts]: this should not be here.

**Comment by mhegazy** [comment]
thanks. one extra statement, seems like a bad merge.

**Comment by mhegazy** [comment]
One failing test now

**Comment by Kingwl** [comment]
everything is ok in my local pc (its second happened)

**Comment by mhegazy** [comment]
i think you also need to accept baselines from exportClassNameWithObjectSystem.js

**Comment by Kingwl** [comment]
@mhegazy  i don't know what happend that all test passed in my mac and windows pc

**Comment by mhegazy** [comment]
you needed to merge master in first. merged manually.

**Comment by Kingwl** [comment]
thanks

**Pull request merged**